datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,820 @@
+            }
+        }
+        Poll::Pending
+    }
+}
+
+pub struct PriorityMap<ID: KeyType, VAL: ValueType> {
+    limit: usize,
+    sort: SortOptions,
+    rnd: RandomState,
+    id_to_val: RawTable<MapItem<ID, VAL>>,
+    val_to_idx: BTreeSet<HeapItem<VAL>>,
+}
+
+pub struct MapItem<ID: KeyType, VAL: ValueType> {
+    hash: u64,
+    pub id: ID,
+    pub val: Option<VAL>,
+}
+
+impl<ID: KeyType, VAL: ValueType> MapItem<ID, VAL> {
+    pub fn new(hash: u64, id: ID, val: Option<VAL>) -> Self {
+        Self { hash, id, val }
+    }
+}
+
+struct HeapItem<VAL: ValueType> {
+    val: Option<VAL>,
+    buk_idx: usize,
+    sort: SortOptions,
+}
+
+impl<VAL: ValueType> HeapItem<VAL> {
+    pub fn new(val: Option<VAL>, buk_idx: usize, sort: SortOptions) -> Self {
+        Self { val, buk_idx, sort }
+    }
+}
+
+impl<VAL: ValueType> Eq for HeapItem<VAL> {}
+
+impl<VAL: ValueType> PartialEq<Self> for HeapItem<VAL> {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl<VAL: ValueType> PartialOrd<Self> for HeapItem<VAL> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<VAL: ValueType> Ord for HeapItem<VAL> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let res = cmp(&self.sort, &self.val, &other.val);
+        if res != Ordering::Equal {
+            return res;
+        }
+        self.buk_idx.cmp(&other.buk_idx)
+    }
+}
+
+impl<ID: KeyType, VAL: ValueType> PriorityMap<ID, VAL>
+where
+    VAL: PartialEq<VAL>,
+{
+    pub fn new(limit: usize, capacity: usize, sort: SortOptions) -> Self {
+        Self {
+            limit,
+            sort,
+            rnd: Default::default(),
+            id_to_val: RawTable::with_capacity(capacity),
+            val_to_idx: Default::default(),
+        }
+    }
+
+    pub fn insert(&mut self, new_id: ID, new_val: Option<VAL>) -> Result<()> {
+        let is_full = self.is_full();
+        assert!(self.len() <= self.limit, "Overflow");
+        let (id_to_val, val_to_idx) = (&mut self.id_to_val, &mut self.val_to_idx);
+
+        // if we're full, and the new val is worse than all our values, just bail
+        if is_full {
+            let worst_entry = val_to_idx.last().expect("Missing value!");
+            if cmp(&self.sort, &new_val, &worst_entry.val) == Ordering::Greater {
+                return Ok(());
+            }
+        }
+
+        // handle new groups we haven't seen yet
+        let mut hasher = self.rnd.build_hasher();
+        new_id.hash(&mut hasher);
+        let new_hash = hasher.finish();
+        let old_bucket = match id_to_val.find(new_hash, |mi| new_id == mi.id) {
+            None => {
+                // we're full and this is a better value, so remove the worst
+                if is_full {
+                    let worst_hi = val_to_idx.pop_last().expect("Missing value!");
+                    unsafe { id_to_val.erase(id_to_val.bucket(worst_hi.buk_idx)) };
```

Comment:
If using the `unsafe` API, we should add some notes that we satisfy the conditions that our usage is safe.

I wonder if the safe `get_mut` API might be good enough in this case 🤔

Reply:
💯 % agree. I was struggling with where and how to best document this though, as I have 8 unsafe blocks in this file.

The reason is that this was a significant performance boost (the ~15%) that got things on par with `main` and should allow the optimizer to be simple and just always use this rule when possible, vs having to do heuristics of some sort.

This optimization takes advantage of the fact that Swiss tables (the implementation in `hashbrown`) don't re-index entries unless the `RawTable` is resized - this is a specialized case that suits us nicely, because we are always accumulating a fixed size (`limit`).

Because the nature of this PR is implementing bi-directionally linked `Map` and `Heap`, this allows us to go from `Heap` -> `Map` with an index directly into the `RawTable`, rather than having to look it up by hash code, then index (which the profiler showed was how that ~15% was being spent).

So in regards to safety, the reason they mark `bucket()` and `bucket_index()` unsafe is because it goes from `index` to memory pointer, and if that `index` is invalid because the `RawTable` resized, then memory corruption could occur. We catch this however by using `try_insert_no_grow()` and catching the error in case the `RawTable` decides to grow, and in that event we rebuild the heap (BTreeSet) so that we always have valid pointers.

This operation is relatively expensive, but we allocate a higher fill factor:

```
            return Ok(Box::new(PrimitiveAggregator::<$vt>::new(
                limit,
                limit * 10,
                sort,
            )))
```

So that this is statistically unlikely. Maybe it will happen on a bad day with some keys that tend to collide, but an exponential resize or two will be amortized away, and even with larger pre-allocated tables (i.e. limit = 1000, 10x) we still use way less memory than when we were storing accumulators for every distinct group prior to this PR.

----

So what do I do with the description above? Just throw it at the top of the file in a doc section? No additional comments needed at the unsafe blocks themselves?

Reply:
Usually we have just a inline comment above the `unsafe` block explaining why it satisfies the safety requirements.

Reply:
Ah, I think [this documentation](https://github.com/apache/arrow-rs/blob/master/arrow/CONTRIBUTING.md#considerations-when-introducing-unsafe) is what I was looking for. I'll try to follow that template.

Reply:
Thanks @Dandandan , I was thinking that documenting every `unsafe {}` block would be burdensome, but doing the exercise helped me as well as the reader regarding clarity. I see why `arrow-rs` put this policy in place!